src/creatures/players/cast/cast_viewer.cpp

```diff
@@ -0,0 +1,1533 @@
+	if (m_owner) {
+		m_owner->sendRestingStatus(protection);
+	}
+}
+
+void CastViewer::AddItem(NetworkMessage &msg, std::shared_ptr<Item> item) {
+	if (m_owner) {
+		m_owner->AddItem(msg, item);
+	}
+}
+
+void CastViewer::AddItem(NetworkMessage &msg, uint16_t id, uint8_t count, uint8_t tier) {
+	if (m_owner) {
+		m_owner->AddItem(msg, id, count, tier);
+	}
+}
+
+void CastViewer::parseSendBosstiary() {
+	if (m_owner) {
+		m_owner->parseSendBosstiary();
+	}
+}
+
+void CastViewer::parseSendBosstiarySlots() {
+	if (m_owner) {
+		m_owner->parseSendBosstiarySlots();
+	}
+}
+
+void CastViewer::sendLootStats(std::shared_ptr<Item> item, uint8_t count) {
+	if (m_owner) {
+		m_owner->sendLootStats(item, count);
+	}
+}
+
+void CastViewer::sendUpdateSupplyTracker(std::shared_ptr<Item> item) {
+	if (m_owner) {
+		m_owner->sendUpdateSupplyTracker(item);
+	}
+}
+
+void CastViewer::sendUpdateImpactTracker(CombatType_t type, int32_t amount) {
+	if (m_owner) {
+		m_owner->sendUpdateImpactTracker(type, amount);
+	}
+}
+
+void CastViewer::openImbuementWindow(std::shared_ptr<Item> item) {
+	if (m_owner) {
+		m_owner->openImbuementWindow(item);
+	}
+}
+
+void CastViewer::sendMarketEnter(uint32_t depotId) {
+	if (m_owner) {
+		m_owner->sendMarketEnter(depotId);
+	}
+}
+
+void CastViewer::sendUnjustifiedPoints(const uint8_t &dayProgress, const uint8_t &dayLeft, const uint8_t &weekProgress, const uint8_t &weekLeft, const uint8_t &monthProgress, const uint8_t &monthLeft, const uint8_t &skullDuration) {
+	if (m_owner) {
+		m_owner->sendUnjustifiedPoints(dayProgress, dayLeft, weekProgress, weekLeft, monthProgress, monthLeft, skullDuration);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendUnjustifiedPoints(dayProgress, dayLeft, weekProgress, weekLeft, monthProgress, monthLeft, skullDuration);
+		}
+	}
+}
+
+void CastViewer::sendModalWindow(const ModalWindow &modalWindow) {
+	if (m_owner) {
+		m_owner->sendModalWindow(modalWindow);
+	}
+}
+
+void CastViewer::sendResourceBalance(Resource_t resourceType, uint64_t value) {
+	if (m_owner) {
+		m_owner->sendResourceBalance(resourceType, value);
+	}
+}
+
+void CastViewer::sendOpenWheelWindow(uint32_t ownerId) {
+	if (m_owner) {
+		m_owner->sendOpenWheelWindow(ownerId);
+	}
+}
+
+void CastViewer::sendCreatureSay(std::shared_ptr<Creature> creature, SpeakClasses type, const std::string &text, const Position* pos) {
+	if (m_owner) {
+		m_owner->sendCreatureSay(creature, type, text, pos);
+
+		if (type == TALKTYPE_PRIVATE_FROM) {
+			return;
+		}
+
+		for (const auto &it : m_viewers) {
+			it.first->sendCreatureSay(creature, type, text, pos);
+		}
+	}
+}
+
+void CastViewer::disconnectClient(const std::string &message) const {
+	if (m_owner) {
+		m_owner->disconnectClient(message);
+	}
+}
+
+void CastViewer::addViewer(ProtocolGame_ptr client, bool spy) {
+	if (m_viewers.find(client) != m_viewers.end()) {
+		return;
+	}
+
+	auto viewerId = m_viewers.size() + 1;
+	std::string guestString = fmt::format("Guest-{}", viewerId);
```

Comment:
I considered this approach, but I have concerns about how it would impact the viewer 'record' feature. Simply incrementing the ID continuously would leave gaps in the sequence when viewers leave, which feels inefficient and messy. Using a per-stream sequence solves the issue of duplicate names, but it doesn't address the problem of maintaining a clean, continuous ID system for current viewers.

Reply:
Why is it inefficient, and why do we care about a continuous range? What am I missing? :D

Reply:
> Why is it inefficient, and why do we care about a continuous range? What am I missing? :D
> 
> Edit: if it's about [9d44a92](https://github.com/opentibiabr/canary/commit/9d44a9231a5c2f5b89a6307005715776216dc647), that uses `m_viewers.size()` right? Shouldn't that be unaffected by the names?

True, I didn't remember that. Thank you.